File: packages/beamtrace_core/src/beamtrace_diff_ffi.mjs
// SPDX-License-Identifier: Apache-2.0 OR MIT

import {
  get as dictGet,
  has as dictHas,
  insert as dictInsert,
  make as makeDict,
} from "../gleam_stdlib/dict.mjs";

const fingerprintModulus = 2147483647;

export function signatureCacheNew() {
  return [];
}

export function signatureCachePut(cache, position, signature) {
  cache[position] = signature;
}

export function signatureCacheGet(cache, position) {
  return cache[position];
}

export function signatureCacheDelete(cache) {
  cache.length = 0;
}

// JavaScript string iteration yields Unicode code points, matching Erlang's
// UTF-8 pattern and the previous Gleam UtfCodepoint implementation.
function extendFingerprint(hash, value) {
  for (const character of value) {
    hash = (hash * 131 + character.codePointAt(0)) % fingerprintModulus;
  }
  return hash;
}

function extendJoined(hash, values) {
  let first = true;
  for (const value of values) {
    if (!first) hash = extendFingerprint(hash, ",");
    hash = extendFingerprint(hash, value);
    first = false;
  }
  return hash;
}

function getOr(dict, key, fallback) {
  return dictHas(dict, key) ? dictGet(dict, key)[0] : fallback;
}

function compactRefined(current, before, after) {
  let hash = extendFingerprint(17, String(current));
  hash = extendFingerprint(hash, "<");
  hash = extendJoined(hash, before.map(String));
  hash = extendFingerprint(hash, ">");
  return extendJoined(hash, after.map(String));
}

function neighbors(id, adjacency, fingerprints) {
  const values = [];
  for (const neighbor of getOr(adjacency, id, [])) {
    values.push(getOr(fingerprints, neighbor, 0));
  }
  values.sort((left, right) => {
    const leftString = String(left);
    const rightString = String(right);
    return leftString < rightString ? -1 : leftString > rightString ? 1 : 0;
  });
  return values;
}

export function compactBase(root, signature) {
  let hash = extendFingerprint(17, root);
  hash = extendFingerprint(hash, "|");
  return extendFingerprint(hash, signature);
}

export function refineRound(fingerprints, incoming, outgoing, ids) {
  let next = makeDict();
  for (const id of ids) {
    next = dictInsert(
      next,
      id,
      compactRefined(
        getOr(fingerprints, id, 0),
        neighbors(id, incoming, fingerprints),
        neighbors(id, outgoing, fingerprints),
      ),
    );
  }
  return next;
}

File: packages/beamtrace_core/src/beamtrace_diff_ffi.erl
%% SPDX-License-Identifier: Apache-2.0 OR MIT
-module(beamtrace_diff_ffi).

-export([
    compact_base/2,
    refine_round/4,
    signature_cache_new/0,
    signature_cache_put/3,
    signature_cache_get/2,
    signature_cache_delete/1
]).

-define(FINGERPRINT_MODULUS, 2147483647).

%% Keep transient alignment strings off the preparing process heap while the
%% four fingerprint rounds allocate their maps. The owning process remains the
%% only reader and ETS also removes the table automatically if that process exits.
signature_cache_new() ->
    ets:new(?MODULE, [set, private]).

signature_cache_put(Cache, Position, Signature) ->
    true = ets:insert(Cache, {Position, Signature}),
    nil.

signature_cache_get(Cache, Position) ->
    ets:lookup_element(Cache, Position, 2).

signature_cache_delete(Cache) ->
    true = ets:delete(Cache),
    nil.

compact_base(Root, Signature) ->
    hash_parts([Root, <<"|">>, Signature], 17).

refine_round(Fingerprints, Incoming, Outgoing, _Ids) ->
    maps:map(fun(Id, Current) ->
        Before = neighbor_fingerprints(Id, Incoming, Fingerprints),
        After = neighbor_fingerprints(Id, Outgoing, Fingerprints),
        compact_refined(Current, Before, After)
    end, Fingerprints).

neighbor_fingerprints(Id, Adjacency, Fingerprints) ->
    case maps:get(Id, Adjacency, []) of
        [] -> [];
        [Neighbor] -> [maps:get(Neighbor, Fingerprints, 0)];
        Neighbors ->
            Values = [maps:get(Neighbor, Fingerprints, 0)
                      || Neighbor <- Neighbors],
            lists:sort(fun lexical_fingerprint/2, Values)
    end.

lexical_fingerprint(Left, Right) ->
    integer_to_binary(Left) =< integer_to_binary(Right).

compact_refined(Current, Before, After) ->
    Hash0 = hash_decimal(Current, 17),
    Hash1 = hash_bytes(<<"<">>, Hash0),
    Hash2 = hash_joined(Before, Hash1),
    Hash3 = hash_bytes(<<">">>, Hash2),
    hash_joined(After, Hash3).

hash_parts([], Hash) -> Hash;
hash_parts([Part | Rest], Hash) ->
    hash_parts(Rest, hash_bytes(Part, Hash)).

hash_joined([], Hash) -> Hash;
hash_joined([First | Rest], Hash) ->
    hash_joined_tail(Rest, hash_decimal(First, Hash)).

hash_joined_tail([], Hash) -> Hash;
hash_joined_tail([Value | Rest], Hash) ->
    Next = hash_decimal(Value, hash_bytes(<<",">>, Hash)),
    hash_joined_tail(Rest, Next).

hash_decimal(Value, Hash) ->
    hash_bytes(integer_to_binary(Value), Hash).

%% Fold UTF-8 code points directly from the binary. This preserves the public
%% fingerprint polynomial while avoiding a code-point list allocation.
hash_bytes(<<>>, Hash) -> Hash;
hash_bytes(<<A, B, C, Rest/binary>>, Hash)
        when A < 128, B < 128, C < 128 ->
    %% Three ASCII code points can be combined with one remainder while the
    %% intermediate stays within a 64-bit Erlang small integer. Canonical
    %% signatures are overwhelmingly ASCII; non-ASCII input falls through to
    %% the exact UTF-8 code-point clause below.
    Next = (
        Hash * 2248091
        + A * 17161
        + B * 131
        + C
    ) rem ?FINGERPRINT_MODULUS,
    hash_bytes(Rest, Next);
hash_bytes(<<Codepoint/utf8, Rest/binary>>, Hash) ->
    Next = (Hash * 131 + Codepoint) rem ?FINGERPRINT_MODULUS,
    hash_bytes(Rest, Next).
